### PR TITLE
Support for Cloud Storage

### LINKED ISSUE #4477: Heroku Amazon s3 “Image source not readable” on upload error
## Version information

 - Laravel: v5.8
 - Voyager: v1.2
 - PHP: v7.2.18
 - Database: Cleardb MySQL (e.g. MySQL 8.0)

## Description

Few days ago I started using laravel with voyager and deployed on heroku server. since heroku uses ephemeral filesystem so I was force to use amazon s3 services. The problem is whenever I upload a image on media manager in my voyager app it gives "Image source not readable" error message but at the same time shows images perfectly. I can only reproduce this error when I'm uploading a image/file or cropping a image. I thought this might be amazon policy issue but I've added policy to allow it still did not work.

**Block public access (bucket settings)**

```
Block all public access
Off
    Block public access to buckets and objects granted through new access control lists (ACLs)
    Off
    Block public access to buckets and objects granted through any access control lists (ACLs)
    Off
    Block public access to buckets and objects granted through new public bucket policies
    Off
    Block public and cross-account access to buckets and objects through any public bucket policies
    Off
```

**Bucket policy editor**

```
{
    "Version": "2012-10-17",
    "Id": "Policy1************",
    "Statement": [
        {
            "Sid": "Stmt1************",
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:*",
            "Resource": "arn:aws:s3:::my_bucket_name/*"
        }
    ]
}
```

## Steps To Reproduce
1. host laravel voyager at heoku and setup amazon s3.
2. Upload Picture in media manager or any file uploader.

## Expected behavior
It should show these errors. it show upload normally like they do in local public folder.

## Screenshots
*Upload Error Image:*

[![Upload Error Image][1]][1]

*Crop Error Image:*

[![Image Crop Error Image][2]][2]

  [1]: https://i.stack.imgur.com/X4ggS.png
  [2]: https://i.stack.imgur.com/zfUW7.png

**EDIT 1**
==========

So I debugged it in vendor directory and turns out it's Intervention\\Image\\Exception\\NotReadableException error in vendor\tcg\voyager\src\Http\Controllers\VoyagerMediaController.php

_vendor\tcg\voyager\src\Http\Controllers\VoyagerMediaController.php_ file

    public function crop(Request $request)
    {
        // Check permission
        $this->authorize('browse_media');

        $createMode = $request->get('createMode') === 'true';
        $x = $request->get('x');
        $y = $request->get('y');
        $height = $request->get('height');
        $width = $request->get('width');

        $realPath = Storage::disk($this->filesystem)->getDriver()->getAdapter()->getPathPrefix();
        $originImagePath = $realPath.$request->upload_path.'/'.$request->originImageName;

        $e = [];
        try {
            if ($createMode) {
                // create a new image with the cpopped data
                $fileNameParts = explode('.', $request->originImageName);
                array_splice($fileNameParts, count($fileNameParts) - 1, 0, 'cropped_'.time());
                $newImageName = implode('.', $fileNameParts);
                $destImagePath = $realPath.$request->upload_path.'/'.$newImageName;
            } else {
                // override the original image
                $destImagePath = $originImagePath;
            }

            Image::make($originImagePath)->crop($width, $height, $x, $y)->save($destImagePath);

            $success = true;
            $message = __('voyager::media.success_crop_image');
        } catch (Exception $e) {
            $success = false;
            $message = $e->getMessage();
            $errors['originImagePath'] = $originImagePath;
            $errors['destImagePath'] = $destImagePath;
            $errors['exception'] = get_class($e);
            $errors['file'] = $e->getFile();
            $errors['code'] = $e->getCode();
            $errors['line'] = $e->getLine();
            $errors['trace'] = $e->getTrace();
        }

        return response()->json(compact('success', 'message', 'errors'));
    }

Post Response _http://localhost/voyager/public/admin/media/crop_

    {
      "success": false,
      "message": "Image source not readable",
      "errors": {
        "originImagePath": "//Capture1.PNG",
        "destImagePath": "//Capture1.cropped_1570987211.PNG",
        "exception": "Intervention\\Image\\Exception\\NotReadableException",
        "file": "D:\\wamp64\\www\\voyager\\vendor\\intervention\\image\\src\\Intervention\\Image\\AbstractDecoder.php",
        "code": 0,
        "line": 346
      }
    }

How do I solve this?

### PR DESCRIPTION
**Pull request by faizananwerali**
Support for Cloud Storage (Google Bucket/Amazon S3) without breaking local and public storage

Fixes https://github.com/the-control-group/voyager/issues/4477

### CODE REVIEW

File: src/Http/Controllers/VoyagerMediaController.php [hunk 1]
```diff
@@ -302,13 +307,16 @@ function ($constraint) use ($thumbnail_data) {
                                 $thumbnail = $this->addWatermarkToImage($thumbnail, $details->watermark);
                             }
                             $thumbnail->save($realPath.$request->upload_path.$name.'-'.($thumbnail_data->name ?? 'thumbnail').'.'.$extension, ($details->quality ?? 90));
+                            //$path = $realPath.$request->upload_path.$name.'-'.($thumbnail_data->name ?? 'thumbnail').'.'.$extension;
```
Comment (fletch3555) [inline src/Http/Controllers/VoyagerMediaController.php]: Please remove these comments.

File: src/Http/Controllers/VoyagerMediaController.php [hunk 2]
```diff
@@ -302,13 +307,16 @@ function ($constraint) use ($thumbnail_data) {
...
+                            //Storage::disk($this->filesystem)->put($path, (string) $thumbnail->encode(null, ($details->quality ?? 90)));
                         }
                     }
                     // Add watermark to image
                     if (property_exists($details, 'watermark') && property_exists($details->watermark, 'source')) {
                         $image = $this->addWatermarkToImage($image, $details->watermark);
                     }
                     $image->save($realPath.$file, ($details->quality ?? 90));
+                    //Storage::disk($this->filesystem)->put($realPath.$file, (string) $image->encode(null, ($details->quality ?? 90)));
```
Comment (fletch3555) [inline src/Http/Controllers/VoyagerMediaController.php]: Same as above

File: src/Http/Controllers/VoyagerMediaController.php [hunk 3]
```diff
@@ -339,34 +347,69 @@ public function crop(Request $request)
...
+            } catch (Exception $e) {
+                $message = $e->getMessage();
+            }
+        }
+
         return response()->json(compact('success', 'message'));
     }
 
-    private function addWatermarkToImage($image, $options)
+    private function addWatermarkToImage(\Intervention\Image\Image $image, $options)
```
Comment (fletch3555) [inline src/Http/Controllers/VoyagerMediaController.php]: `Image` is already imported.  No need to use the full namespace here.

File: src/Http/Controllers/VoyagerMediaController.php [hunk 4]
```diff
@@ -339,34 +347,69 @@ public function crop(Request $request)
...
+        try {
+            $watermark = Image::make(Storage::disk($this->filesystem)->path($options->source));
+            $success = true;
+        } catch (Exception $e) {
+            $success = false;
+        }
+
+        if (!$success) {
+            try {
+                $content = Storage::disk($this->filesystem)->get($options->source);
```
Comment (fletch3555) [inline src/Http/Controllers/VoyagerMediaController.php]: This is the same as line 398 above.  Isn't it guaranteed to fail again?  This whole block seems unnecessary.

**Review by fletch3555 — DISMISSED** [review]

**Comment by MrCrayon** [comment]
Honestly I think this is the wrong approach, upload works so we know we can get the right path that works in all storages without all trial and error
What I would like to see is full path for what works:

- upload
- delete?
- move?
- rename?

And full path for what doesn't work, then we can spot the difference and fix it.

**Comment by faizananwerali** [comment]
Can I give you amazon s3 and Google Bucket Iamuser credentials? Because I
literally don't have anytime this month to debug and both on local and live
server

**Comment by MrCrayon** [comment]
@faizananwerali you can contact me in Voyager Slack Channel

**Comment by faizananwerali** [comment]
Fixed by MrCrayon. https://github.com/MrCrayon/voyager/commit/03f2af479612e20718cacfcf4c5240b2e982049b
Image can't read from cloud so we need to get open the image with Storage and pass the content to Image, inverted process for saving.
Storage should be used without $realPath.
$image->encode()->encoded returns the image as string.

File: src/Http/Controllers/VoyagerMediaController.php (hunk 2 above)
Comment (faizananwerali) [reply src/Http/Controllers/VoyagerMediaController.php]: fixed in latest commit

File: src/Http/Controllers/VoyagerMediaController.php (hunk 1 above)
Comment (faizananwerali) [reply src/Http/Controllers/VoyagerMediaController.php]: fixed in latest commit

File: src/Http/Controllers/VoyagerMediaController.php (hunk 3 above)
Comment (faizananwerali) [reply src/Http/Controllers/VoyagerMediaController.php]: fixed in latest commit

File: src/Http/Controllers/VoyagerMediaController.php (hunk 4 above)
Comment (faizananwerali) [reply src/Http/Controllers/VoyagerMediaController.php]: fixed in latest commit

File: src/Http/Controllers/VoyagerMediaController.php [hunk 5]
```diff
@@ -255,8 +255,8 @@ public function upload(Request $request)
                 'image/svg+xml',
             ];
             if (in_array($request->file->getMimeType(), $imageMimeTypes)) {
-                $image = Image::make($realPath.$file);
-
```
Comment (MrCrayon) [inline src/Http/Controllers/VoyagerMediaController.php:259]: Can you restore the blank lines as the original?
It makes diff much more readable, you can compare with this: https://github.com/MrCrayon/voyager/commit/1a55e5653bfb3d09ce99ef812070bb35fe616d11

File: src/Http/Controllers/VoyagerMediaController.php [hunk 6]
```diff
@@ -338,22 +339,23 @@ public function crop(Request $request)
...
                 $fileNameParts = explode('.', $request->originImageName);
                 array_splice($fileNameParts, count($fileNameParts) - 1, 0, 'cropped_'.time());
                 $newImageName = implode('.', $fileNameParts);
-                $destImagePath = $realPath.$request->upload_path.'/'.$newImageName;
+                $destImagePath = preg_replace('#/+#', '/', $request->upload_path.'/'.$newImageName);
             } else {
                 // override the original image
                 $destImagePath = $originImagePath;
             }
-
```
Comment (MrCrayon) [inline src/Http/Controllers/VoyagerMediaController.php:354]: Same as above

File: src/Http/Controllers/VoyagerMediaController.php [hunk 7]
```diff
@@ -246,7 +246,7 @@ public function upload(Request $request)
             }
 
             $file = $request->file->storeAs($request->upload_path, $name.'.'.$extension, $this->filesystem);
-
```
Comment (MrCrayon) [inline src/Http/Controllers/VoyagerMediaController.php:249]: Same as above

**Review by MrCrayon — CHANGES_REQUESTED** [review]
Please fix the empty lines for clarity.

**Review by MrCrayon — DISMISSED** [review]
Thanks :+1:

**Comment by faizananwerali** [comment]
I don't understand the holdup for merging. Can you please merge it master branch so I don't have to override class

**Comment by faizananwerali** [comment]
what the heck do I have to do to get this pull request approved? I mean I don't understand what do you want more.

**Review by emptynick — APPROVED** [review]

**Comment by emptynick** [comment]
Thanks. But please, watch your tone.
We are all volunteers and have jobs.
It sometimes just takes some time.

**Pull request merged**